{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import vcf\n",
    "import time\n",
    "import re\n",
    "import collections\n",
    "import datetime\n",
    "from scipy.stats import binom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Pileup_line:\n",
    "    def __init__(self, line): \n",
    "        split_line = line.split(\"\\t\")\n",
    "        self.sequence = split_line[0]\n",
    "        self.position = split_line[1]\n",
    "        self.ref_base = split_line[2]\n",
    "        self.read_count = split_line[3]\n",
    "        self.read_results = split_line[4]\n",
    "        self.base_quality = split_line[5]\n",
    "        self.mapping_quality = split_line[6]\n",
    "        \n",
    "    # Pomocna funkcija za racunanje kvaliteta baze iz odgovarajuceg ascii karaktera\n",
    "    # Phred quality score: https://en.wikipedia.org/wiki/Phred_quality_score\n",
    "    def calc_base_quality(ascii_char):\n",
    "        return 1-10**(-ord(ascii_char)/10)\n",
    "    \n",
    "    # Ista funkcija kao gore, samo sto se prosledi redni broj baze/read-a\n",
    "    def calc_base_quality(self, base_num):\n",
    "        # base_num < self.read_count !!!!\n",
    "        return 1-10**(-ord(self.base_quality[base_num])/10)\n",
    "    \n",
    "    def ideal_rr_count(self, read_result):\n",
    "    # read_result je string pa mogu odmah da promijenim , u .\n",
    "    # i mala slova u velika\n",
    "        read_result = read_result.replace(',','.').upper()\n",
    "        char_counter = collections.Counter(read_result).most_common() # Prebroji karaktere i poreda ih od najcescih do najrjedje  \n",
    "        var_list = char_counter \n",
    "        return var_list\n",
    "\n",
    "    # Funkcija za izdvajanjanje insercija ili delecija ili oboje istovremeno, iz originalnog stringa read_result\n",
    "    # Povratna vrijednost su dvije lista, prva lista je lista tacaka, zareza i mozda drugih karaktera\n",
    "    # Druga lista je lista insercija i delecija\n",
    "    def segregate_indels(self, read_result):\n",
    "        # Iskoristila sam re, da na laksi nacin pronadjem brojeve, jer su oni zajednicki za insercije i delecije\n",
    "        # Mozda je ovo moguce napraviti i tako sto se prvo read result iz stringa prebaci u array, ali da se pripazi na visecifrene brojeve\n",
    "        # Pa da se iskoristi np.findall ili slicno da se nadju ti brojevi, i njihovi indeksi i na slican nacin kao u ovom kodu, izbace odgovarajuci el iz tog arraya\n",
    "        numbers = np.array([int(m) for m in re.findall('[\\d]+', read_result)])\n",
    "        # Indeksi brojeva\n",
    "        indices = np.array([m.span() for m in re.finditer('[\\d]+', read_result)])\n",
    "        # String -> np.array\n",
    "        read_result = np.array([char for char in read_result])\n",
    "        # Formiram listu uredjenih dvojki kojih ima koliko i indelsa, a prvi broj u toj dvojki odgovara pocetnom indeksu insercije, tj tacki ili zarezu, \n",
    "        # A drugi broj odgovara indeksu na kom se nalazi zadnji element te insercije\n",
    "        ranges = [(index[0]-2,index[1]+numbers[i]) for i,index in enumerate(indices)]\n",
    "        # Formiranje liste insercija na koju je moguce primjeniti collection.counter\n",
    "        indels_list = []\n",
    "        for item in ranges:\n",
    "            indels_list.append(''.join(read_result[item[0]:item[1]]))\n",
    "\n",
    "        # Brisanje insercija iz read_results\n",
    "        indices_for_removing = []\n",
    "        for item in ranges:\n",
    "            indices_for_removing.extend([*range(item[0],item[1])])\n",
    "        read_result = np.delete(read_result,indices_for_removing)\n",
    "        # read_result to string\n",
    "        read_result = ''.join(read_result[:])\n",
    "        \n",
    "        return read_result, indels_list\n",
    "\n",
    "    def delete_carets(self, read_result):\n",
    "        read_result = re.sub('\\^.','',read_result)\n",
    "        return read_result\n",
    "\n",
    "    def determine_variant(self):\n",
    "        read_result = self.read_results\n",
    "        read_count = self.read_count\n",
    "        # Ovo je prva opcija, znaci da koristim find koja ce samo da posluzi za flag\n",
    "        ins = read_result.find(\"+\") # Flag za inserciju\n",
    "        dele = read_result.find(\"-\") # Flag za deleciju\n",
    "        car = read_result.find(\"^\") # Flag za caret\n",
    "        # Druga opcija je da koristim re bibl da nadjem odmah i indekse svih insercija, delecija i kapica, kako bih onda mogla da ih izbacim\n",
    "        # Zasad ostavljam prvu opciju\n",
    "  \n",
    "        if ins == -1 and dele == -1 and car == -1:\n",
    "            #print(\"No insertion, deletions or carets\")\n",
    "            read_result = re.sub('\\$','',read_result) # PROBAAAA\n",
    "            list_of_vars = self.ideal_rr_count(read_result)\n",
    "            # TO DO\n",
    "            #Varijanta1 = list_of_vars[0]\n",
    "            #Varijanta2 = list_of_vars[1]\n",
    "            Varijanta1 = Variation(list_of_vars[0][0],list_of_vars[0][1], self.read_count)\n",
    "            if len(list_of_vars) > 1:\n",
    "                Varijanta2 = Variation(list_of_vars[1][0],list_of_vars[1][1], self.read_count)\n",
    "            else:\n",
    "                Varijanta2 = None\n",
    "\n",
    "        # Sljedeci slucaj je da nemamo ni insercija ni delecija samo kapicu imamo, izbrisemo kapicu i znak pored, i primjenimo ovo gore\n",
    "        elif (car != -1) and (ins == -1 and dele == -1):\n",
    "            #print(\"Carets yes, no indels\")\n",
    "            #print('Prije brisanja:', read_result)\n",
    "            #print(read_count)\n",
    "            read_result = self.delete_carets(read_result)\n",
    "            read_result = re.sub('\\$','',read_result) # PROBAAAA\n",
    "            # Prakticno smo dobili idealan read result i mozemo pozvati funkciju ideal_rr_count\n",
    "            list_of_vars = self.ideal_rr_count(read_result)\n",
    "            #print('Poslije brisanja:', read_result)\n",
    "            #Varijanta1 = list_of_vars[0]\n",
    "            #Varijanta2 = list_of_vars[1]\n",
    "            Varijanta1 = Variation(list_of_vars[0][0],list_of_vars[0][1], self.read_count)\n",
    "            if len(list_of_vars) > 1:\n",
    "                Varijanta2 = Variation(list_of_vars[1][0],list_of_vars[1][1], self.read_count)\n",
    "            else:\n",
    "                Varijanta2 = None\n",
    "\n",
    "        # Ukoliko imam insercije ili delecije, ili oboje istovremeno\n",
    "        elif (ins != -1) or (dele != -1):\n",
    "            # Provjeri ima li kapica\n",
    "            if car != -1:\n",
    "                #print(\"Carets! Oh, no!!!\")\n",
    "                read_result = self.delete_carets(read_result)\n",
    "            #print(\"No carets\")\n",
    "            read_result = re.sub('\\$','',read_result) # PROBAAAA\n",
    "            read_result, indels_list = self.segregate_indels(read_result)\n",
    "\n",
    "            # zarezi -> tacke, mala slova -> VELIKA SLOVA\n",
    "            indels_list = [s.replace(',','.') for s in indels_list]\n",
    "            number_of_insertions = collections.Counter(map(str.upper, indels_list)).most_common()\n",
    "            # Pozvati ideal_rr_count na read_result\n",
    "            list_of_vars = self.ideal_rr_count(read_result)\n",
    "            list_of_vars.extend(number_of_insertions)\n",
    "            list_of_vars = sorted(list_of_vars, key=lambda tup: tup[1], reverse = True)\n",
    "            #print(self.position, ' ', list_of_vars)\n",
    "\n",
    "            #Varijanta1 = list_of_vars[0]\n",
    "            #Varijanta2 = list_of_vars[1]\n",
    "            Varijanta1 = Variation(list_of_vars[0][0],list_of_vars[0][1], self.read_count)\n",
    "            if len(list_of_vars) > 1:\n",
    "                Varijanta2 = Variation(list_of_vars[1][0],list_of_vars[1][1], self.read_count)\n",
    "            else:\n",
    "                Varijanta2 = None\n",
    "        #print(\"Varijanta1:\", Varijanta1)\n",
    "        #print(\"Varijanta2:\", Varijanta2)\n",
    "\n",
    "        return Varijanta1, Varijanta2\n",
    "            \n",
    "    def determine_genotype(self,variation1, variation2):\n",
    "        genotype_list = []\n",
    "        if variation2 is None:\n",
    "            if self.position == '18027817':\n",
    "                print(1)\n",
    "            genotype_list.append(variation1)\n",
    "            if variation1.variation_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '0/1'\n",
    "            return genotype, genotype_list\n",
    "            \n",
    "        k1 = variation1.number_of_supporting_reads\n",
    "        k2 = variation2.number_of_supporting_reads\n",
    "        n = k1+k2\n",
    "        p = 0.9 # ovaj broj moze da se modifikuje i da predstavlja kvalitet baza\n",
    "        P_k1k1 = binom.pmf(k1,n,p)  # OVDE BILA GRESKA PLAKY\n",
    "        P_k1k2 = binom.pmf(n,n,p)\n",
    "        P_k2k2 = binom.pmf(k2,n,p)\n",
    "        # napomena: brojevi genotipa 1, 2 .. zavise od toga kojim se redosledom pojavljuju u ALT polju u vcf fajlu\n",
    "        # ali cu ja ovde da radim ovako posto mi nista drugacije ne pada na pamet:\n",
    "        # na osnovu onoga sto dobijem ovde za genotip cu da formiram ALT polje\n",
    "        # 1) ako je genotip 0/0 nemamo alt polje uopste\n",
    "        # 2) ako je genotip 1/1, ili 0/1 imam samo jedan slucaj u ALT polju\n",
    "        # 3) ako je genotip 1/2 imacu dva slucaja u ALT polju\n",
    "        # 4) pretpostavljam da necemo imati vise od 3 slucaja u alt polju\n",
    "        if P_k1k1 >= P_k1k2 and P_k1k1 >= P_k2k2: # genotype = 'k1/k1'\n",
    "            if self.position == '18027817':\n",
    "                print('prvi')\n",
    "            if variation1.variation_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '1/1'\n",
    "                genotype_list.append(variation1)\n",
    "        elif P_k1k2 >= P_k1k1 and P_k1k2 >= P_k2k2: # genotype = 'k1/k2'\n",
    "            if self.position == '18027817':\n",
    "                print('drugi')\n",
    "            if variation1.variation_type == 'REF' or variation2.variation_type == 'REF':\n",
    "                genotype = '0/1'\n",
    "                if variation1.variation_type == 'REF':\n",
    "                    genotype_list.append(variation2)\n",
    "                else:\n",
    "                    genotype_list.append(variation1)\n",
    "            else:\n",
    "                genotype = '1/2'\n",
    "                genotype_list.append(variation1)\n",
    "                genotype_list.append(variation2)\n",
    "        #elif P_k1k1 == P_k1k2 or P_k1k1 == P_k2k2 or P_k1k2 == P_k2k2:\n",
    "            \n",
    "        else: # genotype = 'k2/k2'\n",
    "            if self.position == '18027817':\n",
    "                print('treci')\n",
    "            if variation2.variation_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '1/1'\n",
    "                genotype_list.append(variation2)\n",
    "\n",
    "        '''print(self.position, 'variation1:', variation1.variation_type, ' , ', variation1.variation_sequence, ', ', variation1.number_of_supporting_reads, '.......variation2:', variation2.variation_type, ' , ', variation2.variation_sequence, ', ', variation2.number_of_supporting_reads)\n",
    "        print('Pk1k1', P_k1k1, 'Pk1k2', P_k1k2, 'Pk2k2', P_k2k2)\n",
    "        print(genotype, genotype_list[0].variation_sequence)'''\n",
    "        if self.position == '18027817': #29819004\n",
    "            print('var1', variation1.variation_sequence)\n",
    "            print('var2', variation2.variation_sequence)\n",
    "            print('Pk1k1', P_k1k1, 'Pk1k2', P_k1k2, 'Pk2k2', P_k2k2)\n",
    "        return genotype, genotype_list\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Variation:\n",
    "    # variation_type = {'SNV', 'insertion', 'deletion', 'inversion ?'}\n",
    "    # variation_sequence = string od velikih slova, npr 'A' za SNV, 'AG' za inserciju/deleciju\n",
    "    # quality ??\n",
    "    def __init__(self, sequence, number_of_supporting_reads, total_num_of_reads): \n",
    "        self.variation_type, self.variation_sequence, self.variation_len = self.convert_pileup_string(sequence)\n",
    "        self.number_of_supporting_reads = number_of_supporting_reads\n",
    "        self.VAF = float(number_of_supporting_reads)/float(total_num_of_reads)\n",
    "\n",
    "    # Metoda prima string koji je sekvenca preuzeta iz pileup fajla, a vraca dva izlaza na osnovu te sekvence. \n",
    "    # Prvi je variation_type = {'REF', 'SNV', 'insertion', 'deletion'}, a drugi je variation_sequence. Variation_sequence za REF je '.',\n",
    "    # za SNV odgovarajuce veliko slovo npr 'A', a za inserciju i deleciju je string od insertovanih/deletovanih baza npr 'AGC'\n",
    "    def convert_pileup_string(self, sequence):\n",
    "        if len(sequence) == 1:\n",
    "            variation_sequence = sequence\n",
    "            variation_len = 1\n",
    "            if sequence == '.': \n",
    "                variation_type = 'REF'\n",
    "            else:\n",
    "                variation_type = 'SNV'\n",
    "        else:\n",
    "            if sequence[1] == '+':\n",
    "                variation_type = 'insertion'\n",
    "            else:\n",
    "                variation_type = 'deletion'\n",
    "            # kod ispod konvertuje string oblika ,+3AGC u AGC\n",
    "            sequence = sequence[2:]\n",
    "            variation_sequence = ''.join(filter(str.isalpha, sequence))\n",
    "            variation_len = int(''.join(filter(str.isdigit, sequence)))\n",
    "        return variation_type, variation_sequence, variation_len\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "drugi\n",
      "var1 TT\n",
      "var2 T\n",
      "Pk1k1 0.18 Pk1k2 0.81 Pk2k2 0.18\n",
      "186.5842728614807\n"
     ]
    }
   ],
   "source": [
    "with open('/sbgenomics/project-files/merged-normal_pileup.pileup', 'r') as pileup_file, open('vcf4-2.txt', 'w') as vcf:\n",
    "    vcf = create_vcf_header(vcf)\n",
    "    start = time.time()\n",
    "    for line in iter(pileup_file.readline, ''):\n",
    "        position = pileup_file.tell() # uzimamo poziciju dokle smo stigli\n",
    "        pileup = Pileup_line(line)\n",
    "        if pileup.read_count == '0':\n",
    "            genotype = '0/0'\n",
    "        else:\n",
    "            variation1, variation2 = pileup.determine_variant()\n",
    "            genotype, genotype_list = pileup.determine_genotype(variation1, variation2)\n",
    "            \n",
    "        if genotype == '0/0':\n",
    "            continue\n",
    "        elif genotype == '1/1' or genotype == '0/1': # proveri jel / ili \\\n",
    "            variation = genotype_list[0]\n",
    "            VAF = variation.VAF\n",
    "            if variation.variation_type == 'insertion':\n",
    "                REF = pileup.ref_base\n",
    "                ALT = pileup.ref_base + variation.variation_sequence\n",
    "            elif variation.variation_type == 'deletion':\n",
    "                REF = pileup.ref_base\n",
    "                ALT = pileup.ref_base\n",
    "            else: # SNV\n",
    "                REF = pileup.ref_base\n",
    "                ALT = variation.variation_sequence\n",
    "        \n",
    "            if variation.variation_type == 'insertion' or variation.variation_type == 'deletion':\n",
    "                condition = True\n",
    "                del_temp = False # del_temp nam treba samo u slucaju delecije\n",
    "                if variation.variation_type == 'deletion':\n",
    "                    del_temp = True\n",
    "                while(condition):\n",
    "                    for i in range(variation.variation_len):\n",
    "                        line = pileup_file.readline()\n",
    "                        pileup_temp = Pileup_line(line)\n",
    "                        if variation.variation_sequence[i] == pileup_temp.ref_base:\n",
    "                            REF += pileup_temp.ref_base\n",
    "                            del_position = pileup_file.tell()\n",
    "                            if not del_temp:\n",
    "                                ALT += pileup_temp.ref_base\n",
    "                        else:\n",
    "                            condition = False\n",
    "                            if variation.variation_type == 'insertion': \n",
    "                                pileup_file.seek(position)\n",
    "                            #else:\n",
    "                                #pileup_file.seek(del_position)\n",
    "                            break\n",
    "                    del_temp = False\n",
    "        else: # genotype == '1/2' # OVAJ ELSE NIJE ZAVRSEN!!!!\n",
    "            REF = ''\n",
    "            ALT = ''\n",
    "            VAF = ''\n",
    "            for j in range(2):\n",
    "                variation = genotype_list[j]\n",
    "                VAF += '' + str(variation.VAF)\n",
    "                if variation.variation_type == 'insertion':\n",
    "                    REF += pileup.ref_base  # kako se formira uopste ref kad imamo 2 alta?? sta ako je na jednom insercija, a na drugom delecija, ??? promenljiva REF treba da se modifikuje kad skontamo ovo... ALT bi trebalo da je okej\n",
    "                    ALT += pileup.ref_base + variation.variation_sequence\n",
    "                elif variation.variation_type == 'deletion': # OVO TREBA PREPRAVITI\n",
    "                    REF += pileup.ref_base #BEZ VARIAION SEQUENCE\n",
    "                    ALT += pileup.ref_base\n",
    "                else: # SNV\n",
    "                    REF += pileup.ref_base\n",
    "                    ALT += variation.variation_sequence\n",
    "        \n",
    "                if variation.variation_type == 'insertion' or variation.variation_type == 'deletion':\n",
    "                    condition = True\n",
    "                    del_temp = False # del_temp nam treba samo u slucaju delecije\n",
    "                    if variation.variation_type == 'deletion':\n",
    "                        del_temp = True\n",
    "                    while(condition):\n",
    "                        for i in range(variation.variation_len):\n",
    "                            line = pileup_file.readline()\n",
    "                            pileup_temp = Pileup_line(line)\n",
    "                            if variation.variation_sequence[i] == pileup_temp.ref_base:\n",
    "                                REF += pileup_temp.ref_base\n",
    "                                del_position = pileup_file.tell()\n",
    "                                if not del_temp:\n",
    "                                    ALT += pileup_temp.ref_base\n",
    "                            else:\n",
    "                                condition = False\n",
    "                                # ovaj kod ispod je dodat zbog delecije, da ne bismo dodavali suvisne *\n",
    "                                # proveri da li je ovo ok za mikrosatelite! \n",
    "                                if variation.variation_type == 'insertion' or j == 0: \n",
    "                                    pileup_file.seek(position)\n",
    "                                #else:\n",
    "                                    #pileup_file.seek(del_position)\n",
    "                                break\n",
    "                        del_temp = False # ova promenljiva je flag za prvi pass while petlje\n",
    "                if j == 0:\n",
    "                    ALT += ',' \n",
    "                    REF += ','\n",
    "                    VAF += ','\n",
    "        \n",
    "        vcf.write(pileup.sequence + '\\t' + pileup.position + '\\t' + '.' + '\\t' + REF + '\\t' + ALT + '\\t' + 'QUAL' + '\\t' + '.' + '\\t' + '.' + '\\t' + 'GT:VAF' + '\\t' + genotype + ':' + str(VAF) + '\\n')\n",
    "                \n",
    "    end = time.time()\n",
    "    print(end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_vcf_header(vcf):\n",
    "    x = datetime.datetime.now()\n",
    "    fileformat = '##fileformat=VCFv4.2'\n",
    "    date = x.strftime('%Y%m%d')\n",
    "    source = '##source=Jadranka&Suki'\n",
    "    filedate = '##fileDate='+date\n",
    "    reference = '##reference=file:///sbgenomics/Projects/4dc3fffd-51c3-42f3-885b-8b18384be984/human_g1k_v37_decoy.fasta'\n",
    "    vcf.write(fileformat+'\\n' + filedate + '\\n' + source + '\\n' + reference + '\\n')\n",
    "    with open('/sbgenomics/project-files/human_g1k_v37_decoy.fasta.fai','r') as fai:\n",
    "        for line in fai:\n",
    "            split_line = line.split(\"\\t\")\n",
    "            ID = split_line[0]\n",
    "            length = split_line[1]\n",
    "            contig = '##contig=<ID=' + str(ID) + ', length=' + str(length) + '>'\n",
    "            vcf.write(contig + '\\n')\n",
    "    fai.close()\n",
    "    filter = '##FILTER=<ID=PASS,Description=\"All filters passed\">'\n",
    "    vcf.write(filter + '\\n')\n",
    "    vcf.write('#CHROM' + '\\t' + 'POS' + '\\t' + 'ID' + '\\t' + 'REF' + '\\t' + 'ALT' + '\\t' + 'QUAL' + '\\t' + 'FILTER' + '\\t' + 'INFO' + '\\t' + 'FORMAT' + '\\t' + 'HCC1143BL' + '\\n')\n",
    "    return vcf # mozda i ne treba ovo return???"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "230"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int('225')+5"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
