{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import time\n",
    "import re\n",
    "import collections\n",
    "import datetime\n",
    "import math\n",
    "from scipy.stats import binom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Pileup_line:\n",
    "    def __init__(self, line): \n",
    "        split_line = line.split(\"\\t\")\n",
    "        self.sequence = split_line[0]\n",
    "        self.position = split_line[1]\n",
    "        self.ref_base = split_line[2]\n",
    "        self.read_count = split_line[3]\n",
    "        self.read_results = split_line[4]\n",
    "        self.base_quality = split_line[5]\n",
    "        self.mapping_quality = split_line[6]\n",
    "    \n",
    "    # Get sorted list of most common characters from string read_result\n",
    "    def ideal_rr_count(self, read_result):\n",
    "        # Convert ',' to '.' and lower to upper case\n",
    "        read_result = read_result.replace(',','.').upper()\n",
    "        char_counter = collections.Counter(read_result).most_common() \n",
    "        var_list = char_counter \n",
    "        return var_list\n",
    "\n",
    "    # Function for segregating indels from string read_result.\n",
    "    # Output of the function are two lists:\n",
    "    # 1) list of '.', ',', and single chars\n",
    "    # 2) list of insertions or deletions or both\n",
    "    def segregate_indels(self, read_result):\n",
    "        # Find numbers, because both insertion and deletion sequences have them\n",
    "        numbers = np.array([int(m) for m in re.findall('[\\d]+', read_result)])\n",
    "        # Indices of numbers\n",
    "        indices = np.array([m.span() for m in re.finditer('[\\d]+', read_result)])\n",
    "        # String -> np.array\n",
    "        read_result = np.array([char for char in read_result])\n",
    "        # ranges is a list of tuples. Number of tuples is the same as number of INDELs in read_result.\n",
    "        # First number in each tuple is the index of the beginning of the INDEL, which is alway either '.' or ','. Second number is the index of the last element of an INDEL.\n",
    "        ranges = [(index[0]-2,index[1]+numbers[i]) for i,index in enumerate(indices)]\n",
    "        # indels_list is list of indels on which can be applied collection.counter\n",
    "        indels_list = []\n",
    "        for item in ranges:\n",
    "            indels_list.append(''.join(read_result[item[0]:item[1]]))\n",
    "\n",
    "        # Deleting INDELs from read_result\n",
    "        indices_for_removing = []\n",
    "        for item in ranges:\n",
    "            indices_for_removing.extend([*range(item[0],item[1])])\n",
    "        read_result = np.delete(read_result,indices_for_removing)\n",
    "        # read_result to string\n",
    "        read_result = ''.join(read_result[:])\n",
    "        \n",
    "        return read_result, indels_list\n",
    "\n",
    "    # Delete '^' and the following character\n",
    "    def delete_carets(self, read_result):\n",
    "        read_result = re.sub('\\^.','',read_result)\n",
    "        return read_result\n",
    "\n",
    "    # Output of the function are two most common variants in self.read_results.\n",
    "    # Variants are represented with the Variant class.\n",
    "    def determine_variant(self):\n",
    "        read_result = self.read_results\n",
    "        read_count = self.read_count\n",
    "        \n",
    "        ins = read_result.find(\"+\") # insertion flag\n",
    "        dele = read_result.find(\"-\") # deletion flag\n",
    "        car = read_result.find(\"^\") # ^ flag\n",
    "  \n",
    "        # No INDELs\n",
    "        if ins == -1 and dele == -1:\n",
    "            if car != -1:\n",
    "                read_result = self.delete_carets(read_result)\n",
    "            read_result = re.sub('\\$|\\*','',read_result)\n",
    "            list_of_vars = self.ideal_rr_count(read_result)\n",
    "            if len(list_of_vars) == 0:\n",
    "                Variant1 = None\n",
    "                Variant2 = None\n",
    "            elif len(list_of_vars) == 1:\n",
    "                Variant1 = Variant(list_of_vars[0][0],list_of_vars[0][1], self.read_count)\n",
    "                Variant2 = None\n",
    "            else:\n",
    "                Variant1 = Variant(list_of_vars[0][0],list_of_vars[0][1], self.read_count)\n",
    "                Variant2 = Variant(list_of_vars[1][0],list_of_vars[1][1], self.read_count)\n",
    "\n",
    "        # INDELs are present\n",
    "        elif ins != -1 or dele != -1:\n",
    "            if car != -1:\n",
    "                read_result = self.delete_carets(read_result)\n",
    "            read_result = re.sub('\\$|\\*','',read_result)\n",
    "            read_result, indels_list = self.segregate_indels(read_result)\n",
    "\n",
    "            # INDELS_LIST\n",
    "            # ',' -> '.', lower case -> UPPER CASE\n",
    "            indels_list = [s.replace(',','.') for s in indels_list]\n",
    "            number_of_insertions = collections.Counter(map(str.upper, indels_list)).most_common()\n",
    "            # READ_RESULT\n",
    "            list_of_vars = self.ideal_rr_count(read_result)\n",
    "            list_of_vars.extend(number_of_insertions)\n",
    "            list_of_vars = sorted(list_of_vars, key=lambda tup: tup[1], reverse = True)\n",
    "      \n",
    "            if len(list_of_vars) == 0:\n",
    "                Variant1 = None\n",
    "                Variant2 = None\n",
    "            elif len(list_of_vars) == 1:\n",
    "                Variant1 = Variant(list_of_vars[0][0],list_of_vars[0][1], self.read_count)\n",
    "                Variant2 = None\n",
    "            else:\n",
    "                Variant1 = Variant(list_of_vars[0][0],list_of_vars[0][1], self.read_count)\n",
    "                Variant2 = Variant(list_of_vars[1][0],list_of_vars[1][1], self.read_count)\n",
    "            \n",
    "        return Variant1, Variant2\n",
    "            \n",
    "    # Input of determine_genotype are variants outputed from determine_variant().\n",
    "    # Outputs are gonotype and genotype_list.\n",
    "    # Genotype = {'0/0', '0/1', '1/1', '1/2'}\n",
    "    # Genotype_list depends on the genotype; if the genotype is '0/0', genotype_list is empty;\n",
    "    # if the genotype is '0/1' or '1/1' genotype_list consists of a single element, which is\n",
    "    # the corresponding Variant; and if the genotype is '1/2' the lists consists of both variants.\n",
    "    def determine_genotype(self,variant1, variant2):\n",
    "        genotype_list = []\n",
    "        if variant1 is None:\n",
    "            return '0/0', []\n",
    "        if variant2 is None:\n",
    "            if variant1.variant_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '0/1'\n",
    "                genotype_list.append(variant1)\n",
    "            return genotype, genotype_list\n",
    "            \n",
    "        k1 = variant1.number_of_supporting_reads\n",
    "        k2 = variant2.number_of_supporting_reads\n",
    "        n = k1+k2\n",
    "        p = 0.95 # probability of success\n",
    "        P_k1k1 = binom.pmf(k1,n,p)\n",
    "        P_k1k2 = binom.pmf(n,n,p)\n",
    "        P_k2k2 = binom.pmf(k2,n,p)\n",
    "        # Depending on the genotype, the ALT field is later (in the main function) formed:\n",
    "        # 1) 0/0 --> there is no ALT\n",
    "        # 2) 1/1 or 0/1 --> one ALT\n",
    "        # 3) 1/2 --> two ALTs\n",
    "        # The assumption is that there is not going to be more than 2 ALTs\n",
    "        if P_k1k1 >= P_k1k2 and P_k1k1 >= P_k2k2: # genotype = 'k1/k1'\n",
    "            if variant1.variant_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '1/1'\n",
    "                genotype_list.append(variant1)\n",
    "        elif P_k1k2 >= P_k1k1 and P_k1k2 >= P_k2k2: # genotype = 'k1/k2'\n",
    "            if variant1.variant_type == 'REF' or variant2.variant_type == 'REF':\n",
    "                genotype = '0/1'\n",
    "                if variant1.variant_type == 'REF':\n",
    "                    genotype_list.append(variant2)\n",
    "                else:\n",
    "                    genotype_list.append(variant1)\n",
    "            else:\n",
    "                genotype = '1/2'\n",
    "                genotype_list.append(variant1)\n",
    "                genotype_list.append(variant2)\n",
    "        #elif P_k1k1 == P_k1k2 or P_k1k1 == P_k2k2 or P_k1k2 == P_k2k2:   \n",
    "        else: # genotype = 'k2/k2'\n",
    "            if variant2.variant_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '1/1'\n",
    "                genotype_list.append(variant2)\n",
    "\n",
    "        return genotype, genotype_list\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Variant:\n",
    "\n",
    "    def __init__(self, sequence, number_of_supporting_reads, total_num_of_reads): \n",
    "        self.variant_type, self.variant_sequence, self.variant_len = self.convert_pileup_string(sequence)\n",
    "        self.number_of_supporting_reads = number_of_supporting_reads\n",
    "        self.VAF = float(number_of_supporting_reads)/float(total_num_of_reads)\n",
    "\n",
    "    # Input of the function is string for this variant from pileup.\n",
    "    # Outputs of the function are variant_type, variant_sequence and variant_len\n",
    "    # variant_type = {'SNV', 'insertion', 'deletion', 'REF'}\n",
    "    # variant_sequence = '.' for REFs and upper case string for SNVs and INDELS, i.e. 'A' for SNV, 'AG' for INDELs\n",
    "    def convert_pileup_string(self, sequence):\n",
    "        if len(sequence) == 1:\n",
    "            variant_sequence = sequence\n",
    "            variant_len = 1\n",
    "            if sequence == '.': \n",
    "                variant_type = 'REF'\n",
    "            else:\n",
    "                variant_type = 'SNV'\n",
    "        else:\n",
    "            if sequence[1] == '+':\n",
    "                variant_type = 'insertion'\n",
    "            else:\n",
    "                variant_type = 'deletion'\n",
    "            # convert string that looks like this: ,+3AGC  ...to this: AGC\n",
    "            sequence = sequence[2:]\n",
    "            variant_sequence = ''.join(filter(str.isalpha, sequence))\n",
    "            variant_len = int(''.join(filter(str.isdigit, sequence)))\n",
    "        return variant_type, variant_sequence, variant_len\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "191.09894514083862\n"
     ]
    }
   ],
   "source": [
    "with open('/sbgenomics/project-files/merged-normal_pileup.pileup', 'r') as pileup_file, open('vcf4-2garbage.txt', 'w') as vcf:\n",
    "    vcf = create_vcf_header(vcf)\n",
    "    start = time.time()\n",
    "    for line in iter(pileup_file.readline, ''):\n",
    "        position = pileup_file.tell() # take the current cursor position\n",
    "        pileup = Pileup_line(line)\n",
    "        # If there are no reads at this position, we assume that there are no variants\n",
    "        if pileup.read_count == '0':\n",
    "            continue\n",
    "        variant1, variant2 = pileup.determine_variant()\n",
    "        genotype, genotype_list = pileup.determine_genotype(variant1, variant2)\n",
    "        # variants are INDEL --> indel_string = 'INDEL'\n",
    "        # variants are not INDEL --> indel_string = '.'\n",
    "        indel_string = '.'\n",
    "            \n",
    "        # If the genotype is '0/0', there is no variant\n",
    "        if genotype == '0/0':\n",
    "            continue\n",
    "        elif genotype == '1/1' or genotype == '0/1':\n",
    "            variant = genotype_list[0]\n",
    "            VAF = variant.VAF\n",
    "            if variant.variant_type == 'insertion':\n",
    "                REF = pileup.ref_base\n",
    "                ALT = pileup.ref_base + variant.variant_sequence\n",
    "            elif variant.variant_type == 'deletion':\n",
    "                REF = pileup.ref_base\n",
    "                ALT = pileup.ref_base\n",
    "            else: # SNV\n",
    "                REF = pileup.ref_base\n",
    "                ALT = variant.variant_sequence\n",
    "        \n",
    "            if variant.variant_type == 'insertion' or variant.variant_type == 'deletion':\n",
    "                indel_string = 'INDEL'\n",
    "                condition = True\n",
    "                del_temp = False # del_temp variable is needed only in case of deletion\n",
    "                if variant.variant_type == 'deletion':\n",
    "                    del_temp = True\n",
    "                while(condition):\n",
    "                    for i in range(variant.variant_len):\n",
    "                        line = pileup_file.readline()\n",
    "                        pileup_temp = Pileup_line(line)\n",
    "                        if variant.variant_sequence[i] == pileup_temp.ref_base:\n",
    "                            REF += pileup_temp.ref_base\n",
    "                            if not del_temp:\n",
    "                                ALT += pileup_temp.ref_base\n",
    "                        else:\n",
    "                            condition = False\n",
    "                            # The cursor does not go back to variable position in the case of deletion\n",
    "                            # because we want to avoid unnecessary '*'\n",
    "                            if variant.variant_type == 'insertion': \n",
    "                                pileup_file.seek(position)\n",
    "                            break\n",
    "                    del_temp = False\n",
    "            vcf.write(pileup.sequence + '\\t' + pileup.position + '\\t' + '.' + '\\t' + REF + '\\t' + ALT + '\\t' + '.' + '\\t' + '.' + '\\t' + indel_string + '\\t' + 'GT:VAF' + '\\t' + genotype + ':' + str(VAF) + '\\n')\n",
    "        else: # genotype == '1/2'\n",
    "            refs = ['','']\n",
    "            alts = ['','']\n",
    "            vafs = ['','']\n",
    "            indel_string = ['.','.']\n",
    "            for j in range(2):\n",
    "                variant = genotype_list[j]\n",
    "                vafs[j] += str(variant.VAF)\n",
    "                if variant.variant_type == 'insertion':\n",
    "                    refs[j] += pileup.ref_base\n",
    "                    alts[j] += pileup.ref_base + variant.variant_sequence\n",
    "                elif variant.variant_type == 'deletion':\n",
    "                    refs[j] += pileup.ref_base\n",
    "                    alts[j] += pileup.ref_base\n",
    "                else: # SNV\n",
    "                    refs[j] += pileup.ref_base\n",
    "                    alts[j] += variant.variant_sequence\n",
    "        \n",
    "                if variant.variant_type == 'insertion' or variant.variant_type == 'deletion':\n",
    "                    condition = True\n",
    "                    indel_string[j] = 'INDEL'\n",
    "                    del_temp = False # del_temp variable is needed only in case of deletion\n",
    "                    if variant.variant_type == 'deletion':\n",
    "                        del_temp = True\n",
    "                    while(condition):\n",
    "                        for i in range(variant.variant_len):\n",
    "                            line = pileup_file.readline()\n",
    "                            pileup_temp = Pileup_line(line)\n",
    "                            if variant.variant_sequence[i] == pileup_temp.ref_base:\n",
    "                                refs[j] += pileup_temp.ref_base\n",
    "                                if not del_temp:\n",
    "                                    alts[j] += pileup_temp.ref_base\n",
    "                            else:\n",
    "                                condition = False\n",
    "                                # The cursor does not go back to variable position in the case of\n",
    "                                # deletion because we want to avoid unnecessary '*'\n",
    "                                if variant.variant_type == 'insertion' or j == 0: \n",
    "                                    pileup_file.seek(position)\n",
    "                                break\n",
    "                        del_temp = False # flag for first pass of while loop\n",
    "            \n",
    "            if refs[0] == refs[1]:\n",
    "                REF = refs[0]\n",
    "                ALT = alts[0]+','+alts[1]\n",
    "                vcf.write(pileup.sequence + '\\t' + pileup.position + '\\t' + '.' + '\\t' + REF + '\\t' + ALT + '\\t' + '.' + '\\t' + '.' + '\\t' + indel_string[0] + ',' + indel_string[1] + '\\t' + 'GT:VAF' + '\\t' + genotype + ':' + vafs[0]+','+vafs[1] + '\\n')\n",
    "            else:\n",
    "                vcf.write(pileup.sequence + '\\t' + pileup.position + '\\t' + '.' + '\\t' + refs[0] + '\\t' + alts[0] + '\\t' + '.' + '\\t' + '.' + '\\t' + indel_string[0] + '\\t' + 'GT:VAF' + '\\t' + genotype + ':' + vafs[0] + '\\n')\n",
    "                vcf.write(pileup.sequence + '\\t' + pileup.position + '\\t' + '.' + '\\t' + refs[1] + '\\t' + alts[1] + '\\t' + '.' + '\\t' + '.' + '\\t' + indel_string[1] + '\\t' + 'GT:VAF' + '\\t' + genotype + ':' + vafs[1] + '\\n')\n",
    "                \n",
    "    end = time.time()\n",
    "    print(end-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_vcf_header(vcf):\n",
    "    x = datetime.datetime.now()\n",
    "    fileformat = '##fileformat=VCFv4.2'\n",
    "    date = x.strftime('%Y%m%d')\n",
    "    source = '##source=Jadranka&Suki'\n",
    "    filedate = '##fileDate='+date\n",
    "    reference = '##reference=file:///sbgenomics/Projects/4dc3fffd-51c3-42f3-885b-8b18384be984/human_g1k_v37_decoy.fasta'\n",
    "    vcf.write(fileformat+'\\n' + filedate + '\\n' + source + '\\n' + reference + '\\n')\n",
    "    with open('/sbgenomics/project-files/human_g1k_v37_decoy.fasta.fai','r') as fai:\n",
    "        for line in fai:\n",
    "            split_line = line.split(\"\\t\")\n",
    "            ID = split_line[0]\n",
    "            length = split_line[1]\n",
    "            contig = '##contig=<ID=' + str(ID) + ', length=' + str(length) + '>'\n",
    "            vcf.write(contig + '\\n')\n",
    "    fai.close()\n",
    "    filter = '##FILTER=<ID=PASS,Description=\"All filters passed\">'\n",
    "    vcf.write(filter + '\\n')\n",
    "    vcf.write('#CHROM' + '\\t' + 'POS' + '\\t' + 'ID' + '\\t' + 'REF' + '\\t' + 'ALT' + '\\t' + 'QUAL' + '\\t' + 'FILTER' + '\\t' + 'INFO' + '\\t' + 'FORMAT' + '\\t' + 'HCC1143BL' + '\\n')\n",
    "    return vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert pileup.segregate_indels('...+1Aa,,,,-2gc....') == ('..a,,,....', ['.+1A', ',-2gc'])\n",
    "assert pileup.delete_carets('...^|,,^]$..^.') == '...,,$..'\n",
    "assert pileup.ideal_rr_count('...,,,AAGGgg..Tccccc,,,$$**..**atttttttC,,') == [('.', 15), ('T', 8), ('C', 6), ('G', 4), ('*', 4), ('A', 3), ('$', 2)]\n",
    "\n",
    "\n",
    "variant_1 = Variant('A', 2, 10)\n",
    "#variant_2 = Variant(sequence, number_of_supporting_reads, total_num_of_reads)\n",
    "assert variant_1.convert_pileup_string(',+4GCCT') == ('insertion', 'GCCT', 4)\n",
    "assert variant_1.convert_pileup_string(',-1A') == ('deletion','A',1)\n",
    "assert variant_1.convert_pileup_string('.') ==('REF','.',1)\n",
    "assert variant_1.convert_pileup_string('A') == ('SNV','A',1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
