{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/allel/io/vcf_read.py:1516: UserWarning: invalid INFO header: '##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\\n'\n",
      "  warnings.warn('invalid INFO header: %r' % header)\n",
      "/opt/conda/lib/python3.6/site-packages/allel/io/vcf_read.py:1055: UserWarning: 'GT' FORMAT header not found\n",
      "  warnings.warn('%r FORMAT header not found' % name)\n"
     ]
    }
   ],
   "source": [
    "import allel\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv\n",
    "\n",
    "def ref_vcf_to_df():\n",
    "    callset1 = allel.read_vcf('/sbgenomics/project-files/mpileupvcf.mpileup.called.vcf', fields=['variants/REF', 'variants/CHROM', 'variants/POS', 'variants/ALT', 'calldata/GT'])\n",
    "    # Creating ref_vcf DataFrame\n",
    "    df1 = pd.DataFrame.from_dict([callset1['variants/CHROM'], callset1['variants/POS'], callset1['variants/REF']]).T\n",
    "    df1 = df1.rename(columns={0:'CHROM', 1:'POS', 2:'REF'})\n",
    "    df2 = pd.DataFrame.from_dict(callset1['variants/ALT'])\n",
    "    df2 = df2.rename(columns={0:'ALT_1', 1:'ALT_2', 2:'ALT_3'})\n",
    "    df3 = pd.DataFrame.from_dict(callset1['calldata/GT'][:,0])\n",
    "    df3 = df3.rename(columns={0:'GT_1', 1:'GT_2'})\n",
    "    frames = [df1, df2, df3]\n",
    "    ref_vcf = pd.concat(frames, axis=1)\n",
    "    return ref_vcf\n",
    "\n",
    "def our_vcf_to_df(vcf_name):\n",
    "    # Creating our_vcf\n",
    "    callset2 = allel.read_vcf(vcf_name, fields=['variants/REF', 'variants/CHROM', 'variants/POS', 'variants/ALT', 'calldata/GT'])\n",
    "    df1 = pd.DataFrame.from_dict([callset2['variants/CHROM'], callset2['variants/POS'], callset2['variants/REF']]).T\n",
    "    df1 = df1.rename(columns={0:'CHROM', 1:'POS', 2:'REF'})\n",
    "    df2 = pd.DataFrame.from_dict(callset2['variants/ALT'])\n",
    "    df2 = df2.rename(columns={0:'ALT_1', 1:'ALT_2', 2:'ALT_3'})\n",
    "    df3 = pd.DataFrame.from_dict(callset2['calldata/GT'][:,0])\n",
    "    df3 = df3.rename(columns={0:'GT_1', 1:'GT_2'})\n",
    "    frames = [df1, df2, df3]\n",
    "    our_vcf = pd.concat(frames, axis=1)\n",
    "    return our_vcf\n",
    "\n",
    "# Function for creating the list of string, where string represents alleles\n",
    "def allele_list(gt1, gt2, alt1, alt2, alt3, ref):\n",
    "    lst = []\n",
    "    for i in range(0,len(gt1)):\n",
    "        if gt1[i] == 0 and gt2[i] == 0:\n",
    "            lst.append(ref[i])\n",
    "        elif gt1[i] == 0 and gt2[i] == 1:\n",
    "            l = [ref[i],alt1[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt1[i] == 1 and gt2[i] == 0:\n",
    "            l = [alt1[i],ref[i]]\n",
    "            lst.append(''.join(l)) \n",
    "        elif gt1[i] == 1 and gt2[i] == 1:\n",
    "            lst.append(alt1[i])\n",
    "        elif gt1[i] == 0 and gt2[i] == 2:\n",
    "            l = [ref[i],alt2[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt1[i] == 2 and gt2[i] == 0:\n",
    "            l = [alt2[i],ref[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt1[i] == 1 and gt2[i] == 2:\n",
    "            l = [alt1[i], alt2[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt1[i] == 2 and gt2[i] == 1:\n",
    "            l = [alt2[i], alt1[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt1[i] == 2 and gt2[i] == 2: \n",
    "            lst.append(our_alt2[i])\n",
    "        elif gt1[i] == 0 and gt2[i] == 3:\n",
    "            l = [ref[i],alt3[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt1[i] == 3 and gt2[i] == 0:\n",
    "            l = [alt3[i],ref[i]]\n",
    "            lst.append(''.join(l)) \n",
    "        elif gt[i] == 1 and gt2[i] == 3:\n",
    "            l = [alt1[i],alt3[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt[i] == 3 and gt2[i] == 1:\n",
    "            l = [alt3[i],alt1[i]]\n",
    "            lst.append(''.join(l)) \n",
    "        elif gt[i] == 2 and gt2[i] == 3:\n",
    "            l = [alt2[i],alt3[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt[i] == 3 and gt2[i] == 2:\n",
    "            l = [alt3[i],alt2[i]]\n",
    "            lst.append(''.join(l))\n",
    "        elif gt1[i] == 3 and gt2[i] == 3: \n",
    "            lst.append(our_alt3[i])\n",
    "            \n",
    "    return lst\n",
    "\n",
    "def metrics_first(ref_vcf, our_vcf):\n",
    "    # First case:\n",
    "    # TP = number of same positions in our_vcf and ref_vcf with detecred variant (we don't care if variants aren't the same)\n",
    "    # FP = number of those positions in our_vcf which aren't in ref_vcf\n",
    "    # FN = number of those positions in ref_vcf which aren't in our_vcf\n",
    "\n",
    "    # Series in which True means that that position is in ref_vcf and our_vcf at the same time, while False means that some position is in our_vcf but it isn't in ref_vcf\n",
    "    bool_values = our_vcf['POS'].isin(ref_vcf['POS'])\n",
    "    bool_count = bool_values.value_counts()\n",
    "    # Number of trues values in series bool_values is our TP\n",
    "    TP = bool_count[True]\n",
    "    # Number of false values in series bool_values is our FP\n",
    "    FP = bool_count[False]\n",
    "\n",
    "    # # Series in which True means that that position is in ref_vcf and our_vcf at the same time, while False means that some position is in ref_vcf but it isn't in our_vcf\n",
    "    bool_values2 = ref_vcf['POS'].isin(our_vcf['POS'])\n",
    "    bool_count2 = bool_values2.value_counts()\n",
    "    # Number of false values in series bool_values2 is our FN\n",
    "    FN = bool_count2[False]\n",
    "\n",
    "    # Calculating values of metrics\n",
    "    precision = TP/(TP+FP)\n",
    "    recall = TP/(TP+FN)\n",
    "    f1_score =2*precision*recall/(precision+recall)\n",
    "    #print('Metrics in first case:\\nTP:'+str(TP)+'\\nFP:'+str(FP)+'\\nFN:'+str(FN))\n",
    "    #print('Precision:' + str(precision) + '\\nRecall:'+ str(recall) + '\\nF1-score:' + str(f1_score))\n",
    "    return [TP,FP,FN,precision,recall,f1_score]\n",
    "\n",
    "def metrics_second(ref_vcf,our_vcf):\n",
    "    # Second case:\n",
    "    # TP = number of identical variants\n",
    "    # FP - as number of falsely declared variants = number of those positions which are in our_vcf, but aren't in ref_vcf + number of those positions which can be found in both files, but whose variants aren't identical\n",
    "    # FN = number of those positions in ref_vcf which aren't in our_vcf\n",
    "\n",
    "    # Series in which True means that that position is in ref_vcf and our_vcf at the same time, while False means that some position is in our_vcf but it isn't in ref_vcf\n",
    "    bool_values = our_vcf['POS'].isin(ref_vcf['POS'])\n",
    "    bool_count = bool_values.value_counts()\n",
    "    TP_base = bool_count[True] \n",
    "    FP_base = bool_count[False] # number of those positions which are in our_vcf, but aren't in ref_vcf\n",
    "\n",
    "    # List of indices of same positions (positions found in both files) in our_vcf\n",
    "    our_same_pos = bool_values.index[bool_values == True].tolist()\n",
    "\n",
    "    bool_values2 = ref_vcf['POS'].isin(our_vcf['POS'])\n",
    "    bool_count2 = bool_values2.value_counts()\n",
    "    FN_base = bool_count2[False]\n",
    "    # List of indices of same positions (positions found in both files) in ref_vcf\n",
    "    ref_same_pos = bool_values2.index[bool_values2 == True].tolist()\n",
    "    \n",
    "    our_gt1 = our_vcf['GT_1'].iloc[our_same_pos].values\n",
    "    our_gt2 = our_vcf['GT_2'].iloc[our_same_pos].values\n",
    "    our_alt1 = our_vcf['ALT_1'].iloc[our_same_pos].values\n",
    "    our_alt2 = our_vcf['ALT_2'].iloc[our_same_pos].values\n",
    "    our_alt3 = our_vcf['ALT_3'].iloc[our_same_pos].values\n",
    "    our_ref = our_vcf['REF'].iloc[our_same_pos].values\n",
    "    our_lst = allele_list(our_gt1,our_gt2, our_alt1, our_alt2, our_alt3, our_ref)\n",
    "\n",
    "    ref_gt1 = ref_vcf['GT_1'].iloc[ref_same_pos].values\n",
    "    ref_gt2 = ref_vcf['GT_2'].iloc[ref_same_pos].values\n",
    "    ref_alt1 = ref_vcf['ALT_1'].iloc[ref_same_pos].values\n",
    "    ref_alt2 = ref_vcf['ALT_2'].iloc[ref_same_pos].values\n",
    "    ref_alt3 = ref_vcf['ALT_3'].iloc[ref_same_pos].values\n",
    "    ref_ref = ref_vcf['REF'].iloc[ref_same_pos].values\n",
    "    ref_lst = allele_list(ref_gt1, ref_gt2, ref_alt1, ref_alt2, ref_alt3, ref_ref)\n",
    "\n",
    "    our_variants = pd.Series(our_lst)\n",
    "    ref_variants = pd.Series(ref_lst)\n",
    "\n",
    "    # First find identical variants (identical alleles)\n",
    "    var_bool_values = our_variants.isin(ref_variants)\n",
    "    var_bool_count = var_bool_values.value_counts()\n",
    "    TP = var_bool_count[True]\n",
    "    FP_add = var_bool_count[False]\n",
    "    FP = FP_add + FP_base\n",
    "    FN = FN_base\n",
    "    #print('Metrics in second case:\\nTP:'+str(TP)+'\\nFP:'+str(FP)+'\\nFN:'+str(FN))\n",
    "    precision = TP/(TP+FP)\n",
    "    recall = TP/(TP+FN)\n",
    "    f1_score =2*precision*recall/(precision+recall)\n",
    "    #print('Precision:' + str(precision) + '\\nRecall:'+ str(recall) + '\\nF1-score:' + str(f1_score)+'\\n')\n",
    "    return [TP,FP,FN,precision,recall,f1_score]\n",
    "\n",
    "def metrics_third(ref_vcf, our_vcf):\n",
    "    bool_values = our_vcf['POS'].isin(ref_vcf['POS'])\n",
    "    bool_count = bool_values.value_counts()\n",
    "    TP_base = bool_count[True] \n",
    "    FP_base = bool_count[False] # number of those positions which are in our_vcf, but aren't in ref_vcf\n",
    "\n",
    "    # List of indices of same positions (positions found in both files) in our_vcf\n",
    "    our_same_pos = bool_values.index[bool_values == True].tolist()\n",
    "\n",
    "    bool_values2 = ref_vcf['POS'].isin(our_vcf['POS'])\n",
    "    bool_count2 = bool_values2.value_counts()\n",
    "    FN_base = bool_count2[False]\n",
    "    # List of indices of same positions (positions found in both files) in ref_vcf\n",
    "    ref_same_pos = bool_values2.index[bool_values2 == True].tolist()\n",
    "    \n",
    "    our_gt1 = our_vcf['GT_1'].iloc[our_same_pos].values\n",
    "    our_gt2 = our_vcf['GT_2'].iloc[our_same_pos].values\n",
    "    our_alt1 = our_vcf['ALT_1'].iloc[our_same_pos].values\n",
    "    our_alt2 = our_vcf['ALT_2'].iloc[our_same_pos].values\n",
    "    our_alt3 = our_vcf['ALT_3'].iloc[our_same_pos].values\n",
    "    our_ref = our_vcf['REF'].iloc[our_same_pos].values\n",
    "    our_lst = allele_list(our_gt1,our_gt2, our_alt1, our_alt2, our_alt3, our_ref)\n",
    "\n",
    "    ref_gt1 = ref_vcf['GT_1'].iloc[ref_same_pos].values\n",
    "    ref_gt2 = ref_vcf['GT_2'].iloc[ref_same_pos].values\n",
    "    ref_alt1 = ref_vcf['ALT_1'].iloc[ref_same_pos].values\n",
    "    ref_alt2 = ref_vcf['ALT_2'].iloc[ref_same_pos].values\n",
    "    ref_alt3 = ref_vcf['ALT_3'].iloc[ref_same_pos].values\n",
    "    ref_ref = ref_vcf['REF'].iloc[ref_same_pos].values\n",
    "    ref_lst = allele_list(ref_gt1, ref_gt2, ref_alt1, ref_alt2, ref_alt3, ref_ref)\n",
    "\n",
    "    our_variants = pd.Series(our_lst)\n",
    "    ref_variants = pd.Series(ref_lst)\n",
    "\n",
    "    # First find identical variants (identical alleles)\n",
    "    var_bool_values = our_variants.isin(ref_variants)\n",
    "    var_bool_count = var_bool_values.value_counts()\n",
    "    TP = var_bool_count[True]\n",
    "    FN_add = var_bool_count[False]\n",
    "    FN = FN_base + FN_add\n",
    "    FP = FP_base\n",
    "    precision = TP/(TP+FP)\n",
    "    recall = TP/(TP+FN)\n",
    "    f1_score =2*precision*recall/(precision+recall)\n",
    "    return [TP,FP,FN,precision,recall,f1_score]\n",
    "    \n",
    "\n",
    "def main(name):\n",
    "    ref_vcf = ref_vcf_to_df()\n",
    "    ref_len = ref_vcf.shape[0]\n",
    "    our_vcf = our_vcf_to_df(name)\n",
    "    our_len = our_vcf.shape[0]\n",
    "    [TP1, FP1, FN1, precision1, recall1, f1_score1] = metrics_first(ref_vcf, our_vcf)\n",
    "    [TP2, FP2, FN2, precision2, recall2, f1_score2] = metrics_second(ref_vcf, our_vcf)\n",
    "    [TP3, FP3, FN3, precision3, recall3, f1_score3] = metrics_third(ref_vcf, our_vcf)\n",
    "    return [TP1, FP1, FN1, precision1, recall1, f1_score1,TP2, FP2, FN2, precision2, recall2, f1_score2, TP3, FP3, FN3, precision3, recall3, f1_score3, ref_len, our_len]\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    df = pd.DataFrame(columns=['File Name', 'Probability', 'TP1', 'FP1', 'FN1', 'Precision1', 'Recall1', 'F1-score1','TP2', 'FP2', 'FN2', 'Precision2', 'Recall2', 'F1-score2', 'TP3', 'FP3', 'FN3', 'Precision3', 'Recall3', 'F1-score3'])\n",
    "    probabilities = [0.5, 0.55, 0.6, 0.65, 0.70, 0.75, 0.8, 0.85, 0.9, 0.95]\n",
    "    i = 0\n",
    "    for p in probabilities:\n",
    "        name = 'vcf4-2('+str(p)+').txt'\n",
    "        TP1, FP1, FN1, precision1, recall1, f1_score1,TP2, FP2, FN2, precision2, recall2, f1_score2, TP3, FP3, FN3, precision3, recall3, f1_score3, ref_len, our_len = main(name)\n",
    "        df.loc[i] = [name] + [p, TP1, FP1, FN1, precision1, recall1, f1_score1,TP2, FP2, FN2, precision2, recall2, f1_score2, TP3, FP3, FN3, precision3, recall3, f1_score3]\n",
    "        i += 1\n",
    "       \n",
    "    df.to_csv('metrics.csv', index=False)\n",
    "        \n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File Name</th>\n",
       "      <th>Probability</th>\n",
       "      <th>TP1</th>\n",
       "      <th>FP1</th>\n",
       "      <th>FN1</th>\n",
       "      <th>Precision1</th>\n",
       "      <th>Recall1</th>\n",
       "      <th>F1-score1</th>\n",
       "      <th>TP2</th>\n",
       "      <th>FP2</th>\n",
       "      <th>FN2</th>\n",
       "      <th>Precision2</th>\n",
       "      <th>Recall2</th>\n",
       "      <th>F1-score2</th>\n",
       "      <th>TP3</th>\n",
       "      <th>FP3</th>\n",
       "      <th>FN3</th>\n",
       "      <th>Precision3</th>\n",
       "      <th>Recall3</th>\n",
       "      <th>F1-score3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vcf4-2(0.5).txt</td>\n",
       "      <td>0.50</td>\n",
       "      <td>2006</td>\n",
       "      <td>457</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.814454</td>\n",
       "      <td>0.631811</td>\n",
       "      <td>0.711600</td>\n",
       "      <td>1946</td>\n",
       "      <td>517</td>\n",
       "      <td>1169</td>\n",
       "      <td>0.790093</td>\n",
       "      <td>0.624719</td>\n",
       "      <td>0.697741</td>\n",
       "      <td>1946</td>\n",
       "      <td>457</td>\n",
       "      <td>1229</td>\n",
       "      <td>0.809821</td>\n",
       "      <td>0.612913</td>\n",
       "      <td>0.697741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>vcf4-2(0.55).txt</td>\n",
       "      <td>0.55</td>\n",
       "      <td>1959</td>\n",
       "      <td>420</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.823455</td>\n",
       "      <td>0.617008</td>\n",
       "      <td>0.705438</td>\n",
       "      <td>1904</td>\n",
       "      <td>475</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.800336</td>\n",
       "      <td>0.610256</td>\n",
       "      <td>0.692490</td>\n",
       "      <td>1904</td>\n",
       "      <td>420</td>\n",
       "      <td>1271</td>\n",
       "      <td>0.819277</td>\n",
       "      <td>0.599685</td>\n",
       "      <td>0.692490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>vcf4-2(0.6).txt</td>\n",
       "      <td>0.60</td>\n",
       "      <td>1959</td>\n",
       "      <td>420</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.823455</td>\n",
       "      <td>0.617008</td>\n",
       "      <td>0.705438</td>\n",
       "      <td>1904</td>\n",
       "      <td>475</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.800336</td>\n",
       "      <td>0.610256</td>\n",
       "      <td>0.692490</td>\n",
       "      <td>1904</td>\n",
       "      <td>420</td>\n",
       "      <td>1271</td>\n",
       "      <td>0.819277</td>\n",
       "      <td>0.599685</td>\n",
       "      <td>0.692490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>vcf4-2(0.65).txt</td>\n",
       "      <td>0.65</td>\n",
       "      <td>1959</td>\n",
       "      <td>420</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.823455</td>\n",
       "      <td>0.617008</td>\n",
       "      <td>0.705438</td>\n",
       "      <td>1904</td>\n",
       "      <td>475</td>\n",
       "      <td>1216</td>\n",
       "      <td>0.800336</td>\n",
       "      <td>0.610256</td>\n",
       "      <td>0.692490</td>\n",
       "      <td>1904</td>\n",
       "      <td>420</td>\n",
       "      <td>1271</td>\n",
       "      <td>0.819277</td>\n",
       "      <td>0.599685</td>\n",
       "      <td>0.692490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>vcf4-2(0.7).txt</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1977</td>\n",
       "      <td>554</td>\n",
       "      <td>1198</td>\n",
       "      <td>0.781114</td>\n",
       "      <td>0.622677</td>\n",
       "      <td>0.692955</td>\n",
       "      <td>1921</td>\n",
       "      <td>610</td>\n",
       "      <td>1198</td>\n",
       "      <td>0.758989</td>\n",
       "      <td>0.615903</td>\n",
       "      <td>0.680000</td>\n",
       "      <td>1921</td>\n",
       "      <td>554</td>\n",
       "      <td>1254</td>\n",
       "      <td>0.776162</td>\n",
       "      <td>0.605039</td>\n",
       "      <td>0.680000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>vcf4-2(0.75).txt</td>\n",
       "      <td>0.75</td>\n",
       "      <td>2020</td>\n",
       "      <td>843</td>\n",
       "      <td>1155</td>\n",
       "      <td>0.705554</td>\n",
       "      <td>0.636220</td>\n",
       "      <td>0.669096</td>\n",
       "      <td>1965</td>\n",
       "      <td>898</td>\n",
       "      <td>1155</td>\n",
       "      <td>0.686343</td>\n",
       "      <td>0.629808</td>\n",
       "      <td>0.656861</td>\n",
       "      <td>1965</td>\n",
       "      <td>843</td>\n",
       "      <td>1210</td>\n",
       "      <td>0.699786</td>\n",
       "      <td>0.618898</td>\n",
       "      <td>0.656861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>vcf4-2(0.8).txt</td>\n",
       "      <td>0.80</td>\n",
       "      <td>2372</td>\n",
       "      <td>888</td>\n",
       "      <td>804</td>\n",
       "      <td>0.727607</td>\n",
       "      <td>0.746851</td>\n",
       "      <td>0.737104</td>\n",
       "      <td>2325</td>\n",
       "      <td>935</td>\n",
       "      <td>804</td>\n",
       "      <td>0.713190</td>\n",
       "      <td>0.743049</td>\n",
       "      <td>0.727813</td>\n",
       "      <td>2325</td>\n",
       "      <td>888</td>\n",
       "      <td>851</td>\n",
       "      <td>0.723623</td>\n",
       "      <td>0.732053</td>\n",
       "      <td>0.727813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>vcf4-2(0.85).txt</td>\n",
       "      <td>0.85</td>\n",
       "      <td>2932</td>\n",
       "      <td>1532</td>\n",
       "      <td>244</td>\n",
       "      <td>0.656810</td>\n",
       "      <td>0.923174</td>\n",
       "      <td>0.767539</td>\n",
       "      <td>2888</td>\n",
       "      <td>1576</td>\n",
       "      <td>244</td>\n",
       "      <td>0.646953</td>\n",
       "      <td>0.922095</td>\n",
       "      <td>0.760400</td>\n",
       "      <td>2888</td>\n",
       "      <td>1532</td>\n",
       "      <td>288</td>\n",
       "      <td>0.653394</td>\n",
       "      <td>0.909320</td>\n",
       "      <td>0.760400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>vcf4-2(0.9).txt</td>\n",
       "      <td>0.90</td>\n",
       "      <td>3116</td>\n",
       "      <td>3027</td>\n",
       "      <td>60</td>\n",
       "      <td>0.507244</td>\n",
       "      <td>0.981108</td>\n",
       "      <td>0.668741</td>\n",
       "      <td>3076</td>\n",
       "      <td>3067</td>\n",
       "      <td>60</td>\n",
       "      <td>0.500733</td>\n",
       "      <td>0.980867</td>\n",
       "      <td>0.663002</td>\n",
       "      <td>3076</td>\n",
       "      <td>3027</td>\n",
       "      <td>100</td>\n",
       "      <td>0.504014</td>\n",
       "      <td>0.968514</td>\n",
       "      <td>0.663002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>vcf4-2(0.95).txt</td>\n",
       "      <td>0.95</td>\n",
       "      <td>3165</td>\n",
       "      <td>6769</td>\n",
       "      <td>12</td>\n",
       "      <td>0.318603</td>\n",
       "      <td>0.996223</td>\n",
       "      <td>0.482801</td>\n",
       "      <td>3123</td>\n",
       "      <td>6811</td>\n",
       "      <td>12</td>\n",
       "      <td>0.314375</td>\n",
       "      <td>0.996172</td>\n",
       "      <td>0.477925</td>\n",
       "      <td>3123</td>\n",
       "      <td>6769</td>\n",
       "      <td>54</td>\n",
       "      <td>0.315710</td>\n",
       "      <td>0.983003</td>\n",
       "      <td>0.477925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          File Name  Probability   TP1   FP1   FN1  Precision1   Recall1  \\\n",
       "0   vcf4-2(0.5).txt         0.50  2006   457  1169    0.814454  0.631811   \n",
       "1  vcf4-2(0.55).txt         0.55  1959   420  1216    0.823455  0.617008   \n",
       "2   vcf4-2(0.6).txt         0.60  1959   420  1216    0.823455  0.617008   \n",
       "3  vcf4-2(0.65).txt         0.65  1959   420  1216    0.823455  0.617008   \n",
       "4   vcf4-2(0.7).txt         0.70  1977   554  1198    0.781114  0.622677   \n",
       "5  vcf4-2(0.75).txt         0.75  2020   843  1155    0.705554  0.636220   \n",
       "6   vcf4-2(0.8).txt         0.80  2372   888   804    0.727607  0.746851   \n",
       "7  vcf4-2(0.85).txt         0.85  2932  1532   244    0.656810  0.923174   \n",
       "8   vcf4-2(0.9).txt         0.90  3116  3027    60    0.507244  0.981108   \n",
       "9  vcf4-2(0.95).txt         0.95  3165  6769    12    0.318603  0.996223   \n",
       "\n",
       "   F1-score1   TP2   FP2   FN2  Precision2   Recall2  F1-score2   TP3   FP3  \\\n",
       "0   0.711600  1946   517  1169    0.790093  0.624719   0.697741  1946   457   \n",
       "1   0.705438  1904   475  1216    0.800336  0.610256   0.692490  1904   420   \n",
       "2   0.705438  1904   475  1216    0.800336  0.610256   0.692490  1904   420   \n",
       "3   0.705438  1904   475  1216    0.800336  0.610256   0.692490  1904   420   \n",
       "4   0.692955  1921   610  1198    0.758989  0.615903   0.680000  1921   554   \n",
       "5   0.669096  1965   898  1155    0.686343  0.629808   0.656861  1965   843   \n",
       "6   0.737104  2325   935   804    0.713190  0.743049   0.727813  2325   888   \n",
       "7   0.767539  2888  1576   244    0.646953  0.922095   0.760400  2888  1532   \n",
       "8   0.668741  3076  3067    60    0.500733  0.980867   0.663002  3076  3027   \n",
       "9   0.482801  3123  6811    12    0.314375  0.996172   0.477925  3123  6769   \n",
       "\n",
       "    FN3  Precision3   Recall3  F1-score3  \n",
       "0  1229    0.809821  0.612913   0.697741  \n",
       "1  1271    0.819277  0.599685   0.692490  \n",
       "2  1271    0.819277  0.599685   0.692490  \n",
       "3  1271    0.819277  0.599685   0.692490  \n",
       "4  1254    0.776162  0.605039   0.680000  \n",
       "5  1210    0.699786  0.618898   0.656861  \n",
       "6   851    0.723623  0.732053   0.727813  \n",
       "7   288    0.653394  0.909320   0.760400  \n",
       "8   100    0.504014  0.968514   0.663002  \n",
       "9    54    0.315710  0.983003   0.477925  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
