{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import vcf\n",
    "import time\n",
    "import re\n",
    "from scipy.stats import binom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Pileup_line:\n",
    "    def __init__(self, line): \n",
    "        split_line = line.split(\"\\t\")\n",
    "        self.sequence = split_line[0]\n",
    "        self.position = split_line[1]\n",
    "        self.ref_base = split_line[2]\n",
    "        self.read_count = split_line[3]\n",
    "        self.read_results = split_line[4]\n",
    "        self.base_quality = split_line[5]\n",
    "        self.mapping_quality = split_line[6]\n",
    "        \n",
    "    # Pomocna funkcija za racunanje kvaliteta baze iz odgovarajuceg ascii karaktera\n",
    "    # Phred quality score: https://en.wikipedia.org/wiki/Phred_quality_score\n",
    "    def calc_base_quality(ascii_char):\n",
    "        return 1-10**(-ord(ascii_char)/10)\n",
    "    \n",
    "    # Ista funkcija kao gore, samo sto se prosledi redni broj baze/read-a\n",
    "    def calc_base_quality(self, base_num):\n",
    "        # base_num < self.read_count !!!!\n",
    "        return 1-10**(-ord(self.base_quality[base_num])/10)\n",
    "    \n",
    "    # Pomocna funkcija za racunanje broja pojavljivanja karaktera u stringu\n",
    "    # Ovo sam mislila za prebrojavanje npr baza koje se poklapaju sa referentnom tj. broj . u read_results\n",
    "    def num_of_occurence(self, char):\n",
    "        return self.read_results.count(char)\n",
    "    \n",
    "    def count(self):\n",
    "        #treba resiti pocetak i kraj stringa tj ^neki_znak i neki_znak$ ignorisati\n",
    "        \n",
    "        #sa ovim ispod izvrsavanje koda traje 28sekundi\n",
    "        #posle ^ moze da ide bilo koji karakter, ukljucujuci i one bitne pa sa ovom linijom to brisem da ne bismo pogresno racunale\n",
    "        #self.read_results = re.sub(r\"\\^.\",'',self.read_results) \n",
    "\n",
    "        \n",
    "        num_C = self.read_results.count('g') + self.read_results.count('C')\n",
    "        num_G = self.read_results.count('c') + self.read_results.count('G')\n",
    "        num_A = self.read_results.count('t') + self.read_results.count('A')\n",
    "        num_T = self.read_results.count('a') + self.read_results.count('T')\n",
    "        num_match = self.read_results.count('.') + self.read_results.count(',')\n",
    "        num_deletions = self.read_results.count('*') + self.read_results.count('-')\n",
    "        num_insertions = self.read_results.count('+')\n",
    "        \n",
    "        if self.ref_base == 'C':\n",
    "            num_C += num_match\n",
    "        elif self.ref_base == 'G':\n",
    "            num_G += num_match\n",
    "        elif self.ref_base == 'A':\n",
    "            num_A += num_match\n",
    "        elif self.ref_base == 'T':\n",
    "            num_T += num_match\n",
    "            \n",
    "    def determine_genotype(variation1, variation2):\n",
    "        k1 = variation1.number_of_supporting_reads\n",
    "        k2 = variation2.number_of_supporting_reads\n",
    "        n = k1+k2\n",
    "        p = 0.5 # ovaj broj moze da se modifikuje i da predstavlja kvalitet baza\n",
    "        P_k1k1 = binom.pmf(k1,n,p)\n",
    "        P_k1k2 = binom.pmf(k2,n,p)\n",
    "        P_k2k2 = binom.pmf(n,n,p)\n",
    "        # napomena: brojevi genotipa 1, 2 .. zavise od toga kojim se redosledom pojavljuju u ALT polju u vcf fajlu\n",
    "        # ali cu ja ovde da radim ovako posto mi nista drugacije ne pada na pamet:\n",
    "        # na osnovu onoga sto dobijem ovde za genotip cu da formiram ALT polje\n",
    "        # 1) ako je genotip 0/0 nemamo alt polje uopste\n",
    "        # 2) ako je genotip 1/1, ili 0/1 imam samo jedan slucaj u ALT polju\n",
    "        # 3) ako je genotip 1/2 imacu dva slucaja u ALT polju\n",
    "        # 4) pretpostavljam da necemo imati vise od 3 slucaja u alt polju\n",
    "        genotype_list = []\n",
    "        if P_k1k1 > P_k1k2 and P_k1k1 > P_k2k2: # genotype = 'k1/k1'\n",
    "            if variation1.variation_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '1/1'\n",
    "                genotype_list.append(variation1)\n",
    "        elif P_k1k2 > P_k1k1 and P_k1k2 > P_k2k2: # genotype = 'k1/k2'\n",
    "            if variation1.variation_type == 'REF' or variation2.variation_type == 'REF':\n",
    "                genotype = '0/1'\n",
    "                if variation1.variation_type == 'REF':\n",
    "                    genotype_list.append(variation1)\n",
    "                else:\n",
    "                    genotype_list.append(variation2)\n",
    "            else:\n",
    "                genotype = '1/2'\n",
    "                genotype_list.append(variation1)\n",
    "                genotype_list.append(variation2)\n",
    "        else: # genotype = 'k2/k2'\n",
    "            if variation2.variation_type == 'REF':\n",
    "                genotype = '0/0'\n",
    "            else:\n",
    "                genotype = '1/1'\n",
    "                genotype_list.append(variation2)\n",
    "\n",
    "        return genotype, genotype_list\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Variation:\n",
    "    # variation_type = {'SNV', 'insertion', 'deletion', 'inversion ?'}\n",
    "    # variation_sequence = string od velikih slova, npr 'A' za SNV, 'AG' za inserciju/deleciju\n",
    "    # quality ??\n",
    "    def __init__(self, sequence, num_of_reads): \n",
    "        self.variation_type, self.variation_sequence = self.convert_pileup_string(sequence)\n",
    "        self.number_of_supporting_reads = num_of_reads\n",
    "\n",
    "    # Metoda prima string koji je sekvenca preuzeta iz pileup fajla, a vraca dva izlaza na osnovu te sekvence. \n",
    "    # Prvi je variation_type = {'REF', 'SNV', 'insertion', 'deletion'}, a drugi je variation_sequence. Variation_sequence za REF je '.',\n",
    "    # za SNV odgovarajuce veliko slovo npr 'A', a za inserciju i deleciju je string od insertovanih/deletovanih baza npr 'AGC'\n",
    "    def convert_pileup_string(sequence):\n",
    "        if len(sequence) == 1:\n",
    "            if sequence == '.' or sequence == ',': # da li ima potrebe za or sequence == ',' , tj. da li se i , prosledjuje kao .?\n",
    "                variation_type = 'REF'\n",
    "                variation_sequence = '.'\n",
    "            else:\n",
    "                variation_type = 'SNV'\n",
    "                variation_sequence = sequence.upper() # da li ima potrebe za ovim upper, tj. da li se svi SNV prosledjuju kao velika slova?\n",
    "        else:\n",
    "            if sequence[1] == '+':\n",
    "                variation_type = 'insertion'\n",
    "            else:\n",
    "                variation_type = 'deletion'\n",
    "            # kod ispod konvertuje string oblika ,+3agc u AGC\n",
    "            sequence = sequence[2:]\n",
    "            sequence = sequence.upper()\n",
    "            variation_sequence = ''.join(filter(str.isalpha, sequence))\n",
    "            return variation_type, variation_sequence\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28.33063554763794\n"
     ]
    }
   ],
   "source": [
    "with open('/sbgenomics/project-files/merged-normal_pileup.pileup', 'r') as pileup_file:\n",
    "    start = time.time()\n",
    "    for line in iter(pileup_file.readline, ''):\n",
    "        position = pileup_file.tell() # uzimamo poziciju dokle smo stigli\n",
    "        pileup = Pileup_line(line)\n",
    "        variation1, variation2 = pileup.determine_variation()\n",
    "        genotype, genotype_list = pileup.determine_genotype(variation1, variation2)\n",
    "        if genotype == '0/0':\n",
    "            continue\n",
    "        elif genotype == '1/1' or genotype == '0/1': # proveri jel / ili \\\n",
    "            variation = genotype_list[0] \n",
    "            if variation.variation_type == 'insertion':\n",
    "                REF = pileup.ref_base\n",
    "                ALT = pileup.ref_base + variation.variation_sequence\n",
    "            elif variation_type == 'deletion':\n",
    "                REF = pileup.ref_base + variation.variation_sequence\n",
    "                ALT = pileup.ref_base\n",
    "            else: # SNV\n",
    "                REF = pileup.ref_base\n",
    "                ALT = variation.variation_sequence\n",
    "        \n",
    "            if variation.variation_type == 'insertion' or variation.variation_type == 'deletion':\n",
    "                condition = True\n",
    "                while(condition):\n",
    "                    for i in range(len(variation.variation_sequence)):\n",
    "                        line = pileup_file.readline()\n",
    "                        pileup = Pileup_line(line)\n",
    "                        if variation.variation_sequence[i] == pileup.ref_base:\n",
    "                            ALT += ref_base\n",
    "                            REF += ref_base\n",
    "                        else:\n",
    "                            condition = false\n",
    "                            pileup_file.seek(position)\n",
    "                            break\n",
    "        else: # genotype == '1/2' # OVAJ ELSE NIJE ZAVRSEN!!!!\n",
    "            REF = ''\n",
    "            ALT = ''\n",
    "            for i in range(2):\n",
    "                variation = genotype_list[i] \n",
    "                if variation.variation_type == 'insertion':\n",
    "                    REF += pileup.ref_base  # kako se formira uopste ref kad imamo 2 alta?? sta ako je na jednom insercija, a na drugom delecija, ??? promenljiva REF treba da se modifikuje kad skontamo ovo... ALT bi trebalo da je okej\n",
    "                    ALT += pileup.ref_base + variation.variation_sequence\n",
    "                elif variation_type == 'deletion':\n",
    "                    REF += pileup.ref_base + variation.variation_sequence\n",
    "                    ALT += pileup.ref_base\n",
    "                else: # SNV\n",
    "                    REF += pileup.ref_base\n",
    "                    ALT += variation.variation_sequence\n",
    "        \n",
    "                if variation.variation_type == 'insertion' or variation.variation_type == 'deletion':\n",
    "                    condition = True\n",
    "                    while(condition):\n",
    "                        for i in range(len(variation.variation_sequence)):\n",
    "                            line = pileup_file.readline()\n",
    "                            pileup = Pileup_line(line)\n",
    "                            if variation.variation_sequence[i] == pileup.ref_base:\n",
    "                                ALT += ref_base\n",
    "                                REF += ref_base\n",
    "                            else:\n",
    "                                condition = false\n",
    "                                pileup_file.seek(position)\n",
    "                                break\n",
    "                if i == 0:\n",
    "                    ALT += ','\n",
    "                \n",
    "\n",
    "\n",
    "    end = time.time()\n",
    "    print(end-start)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
