{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.6.3"
    },
    "colab": {
      "name": "Copy of skripta.ipynb",
      "provenance": [],
      "collapsed_sections": []
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "ucsWcq9e6xd4",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import vcf\n",
        "import time\n",
        "import re\n",
        "from scipy.stats.binom import pmf"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "peFoRlgP6xeE",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "class Pileup_line:\n",
        "    def __init__(self, line): \n",
        "        split_line = line.split(\"\\t\")\n",
        "        self.sequence = split_line[0]\n",
        "        self.position = split_line[1]\n",
        "        self.ref_base = split_line[2]\n",
        "        self.read_count = split_line[3]\n",
        "        self.read_results = split_line[4]\n",
        "        self.base_quality = split_line[5]\n",
        "        self.mapping_quality = split_line[6]\n",
        "        \n",
        "    # Pomocna funkcija za racunanje kvaliteta baze iz odgovarajuceg ascii karaktera\n",
        "    # Phred quality score: https://en.wikipedia.org/wiki/Phred_quality_score\n",
        "    def calc_base_quality(ascii_char):\n",
        "        return 1-10**(-ord(ascii_char)/10)\n",
        "    \n",
        "    # Ista funkcija kao gore, samo sto se prosledi redni broj baze/read-a\n",
        "    def calc_base_quality(self, base_num):\n",
        "        # base_num < self.read_count !!!!\n",
        "        return 1-10**(-ord(self.base_quality[base_num])/10)\n",
        "    \n",
        "    # Pomocna funkcija za racunanje broja pojavljivanja karaktera u stringu\n",
        "    # Ovo sam mislila za prebrojavanje npr baza koje se poklapaju sa referentnom tj. broj . u read_results\n",
        "    def num_of_occurence(self, char):\n",
        "        return self.read_results.count(char)\n",
        "    \n",
        "    def count(self):\n",
        "        #treba resiti pocetak i kraj stringa tj ^neki_znak i neki_znak$ ignorisati\n",
        "        \n",
        "        #sa ovim ispod izvrsavanje koda traje 28sekundi\n",
        "        #posle ^ moze da ide bilo koji karakter, ukljucujuci i one bitne pa sa ovom linijom to brisem da ne bismo pogresno racunale\n",
        "        #self.read_results = re.sub(r\"\\^.\",'',self.read_results) \n",
        "\n",
        "        \n",
        "        num_C = self.read_results.count('g') + self.read_results.count('C')\n",
        "        num_G = self.read_results.count('c') + self.read_results.count('G')\n",
        "        num_A = self.read_results.count('t') + self.read_results.count('A')\n",
        "        num_T = self.read_results.count('a') + self.read_results.count('T')\n",
        "        num_match = self.read_results.count('.') + self.read_results.count(',')\n",
        "        num_deletions = self.read_results.count('*') + self.read_results.count('-')\n",
        "        num_insertions = self.read_results.count('+')\n",
        "        \n",
        "        if self.ref_base == 'C':\n",
        "            num_C += num_match\n",
        "        elif self.ref_base == 'G':\n",
        "            num_G += num_match\n",
        "        elif self.ref_base == 'A':\n",
        "            num_A += num_match\n",
        "        elif self.ref_base == 'T':\n",
        "            num_T += num_match\n",
        "        \n",
        "\n",
        "    def determine_genotype(Variation variation1, Variation variation2):\n",
        "\t\t    k1 = variation1.number_of_supporting_reads\n",
        "\t\t    k2 = variation2.number_of_supporting_reads\n",
        "\t\t    n = k1+k2\n",
        "\t\t    p = 0.5 # ovaj broj moze da se modifikuje i da predstavlja kvalitet baza\n",
        "\t\t    P_k1k1 = pmf(k1,n,p)\n",
        "\t\t    P_k1k2 = pmf(k2,n,p)\n",
        "\t\t    P_k2k2 = pmf(n,n,p)\n",
        "\t\t    # napomena: brojevi genotipa 1, 2 .. zavise od toga kojim se redosledom pojavljuju u ALT polju u vcf fajlu\n",
        "\t\t    # ali cu ja ovde da radim ovako posto mi nista drugacije ne pada na pamet:\n",
        "\t\t    # na osnovu onoga sto dobijem ovde za genotip cu da formiram ALT polje\n",
        "\t\t    # 1) ako je genotip 0/0 nemamo alt polje uopste\n",
        "\t\t    # 2) ako je genotip 1/1, ili 0/1 imam samo jedan slucaj u ALT polju\n",
        "\t\t    # 3) ako je genotip 1/2 imacu dva slucaja u ALT polju\n",
        "\t\t    # 4) pretpostavljam da necemo imati vise od 3 slucaja u alt polju\n",
        "\t\t    genotype_list = []\n",
        "\t\t    if P_k1k1 > P_k1k2 and P_k1k1 > P_k2k2: # genotype = 'k1/k1'\n",
        "\t\t\t      if variation1.variation_type == 'REF':\n",
        "\t\t\t\t        genotype = '0/0'\n",
        "\t\t\t      else:\n",
        "\t\t\t\t        genotype = '1/1'\n",
        "\t\t\t\t        genotype_list.append(variation1)\n",
        "\t\t    elif P_k1k2 > P_k1k1 and P_k1k2 > P_k2k2: # genotype = 'k1/k2'\n",
        "\t\t\t      if variation1.variation_type == 'REF' or variation2.variation_type == 'REF':\n",
        "\t\t\t\t        genotype = '0/1'\n",
        "\t\t\t\t        if variation1.variation_type == 'REF':\n",
        "\t\t\t\t\t          genotype_list.append(variation1)\n",
        "\t\t\t\t        else:\n",
        "\t\t\t\t\t          genotype_list.append(variation2)\n",
        "\t\t\t      else:\n",
        "\t\t\t\t        genotype = '1/2'\n",
        "\t\t\t\t        genotype_list.append(variation1)\n",
        "\t\t\t\t        genotype_list.append(variation2)\n",
        "\t\t    else: # genotype = 'k2/k2'\n",
        "\t\t\t      if variation2.variation_type == 'REF':\n",
        "\t\t\t\t        genotype = '0/0'\n",
        "\t\t\t      else:\n",
        "\t\t\t\t        genotype = '1/1'\n",
        "\t\t\t\t        genotype_list.append(variation2)\n",
        "\t\t\n",
        "\t\t    return genotype, genotype_list\n",
        "\n"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "U9UEiFsD6xeO",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "class Variation: # ili tako nesto\n",
        "    # variation_type = {'SNV', 'insertion', 'deletion', 'inversion ?'}\n",
        "    # variation_sequence = string od velikih slova, npr 'A' za SNV, 'AG' za inserciju/deleciju\n",
        "    # quality ??\n",
        "    \n",
        "    # prosta metoda koja ce da konvertuje string iz pileupa: ,+1ga --> GA"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "rdOmVmFH6xeb",
        "colab_type": "code",
        "outputId": "4870e0ca-24c7-49d1-82fe-56087026293f",
        "colab": {}
      },
      "source": [
        "with open('/sbgenomics/project-files/merged-normal_pileup.pileup', 'r') as pileup_file:\n",
        "    start = time.time()\n",
        "    for line in iter(pileup_file.readline, ''):\n",
        "        position = pileup_file.tell() # uzimamo poziciju dokle smo stigli\n",
        "        pileup = Pileup_line(line)\n",
        "\t\t    variation1, variation2 = pileup.determine_variation()\n",
        "        genotype, genotype_list = pileup.determine_genotype(variation1, variation2)\n",
        "        #inicijalizacija polja ref i alt u zavisnosti od tipa varijacije\n",
        "        if genotype == '0/0':\n",
        "\t\t\t      continue\n",
        "\t\t    elif genotype == '1/1' or genotype == '0/1': # proveri jel / ili \\\n",
        "\t\t\t      variation = genotype_list[0] # u ovim slucajevima imamo samo jedan element u genotype_list.. ovako je samo lakse dalje\n",
        "            if variation.variation_type == 'insertion':\n",
        "                REF = pileup.ref_base\n",
        "                ALT = pileup.ref_base + variation.variation_sequence\n",
        "            elif variation_type == 'deletion':\n",
        "                REF = pileup.ref_base + variation.variation_sequence\n",
        "                ALT = pileup.ref_base\n",
        "            else: # SNV\n",
        "                REF = pileup.ref_base\n",
        "                ALT = variation.variation_sequence\n",
        "        \n",
        "            if variation.variation_type == 'insertion' or variation.variation_type == 'deletion':\n",
        "                condition = True\n",
        "                while(condition):\n",
        "                    for i in range(len(variation.variation_sequence)):\n",
        "                        line = pileup_file.readline()\n",
        "                        pileup = Pileup_line(line)\n",
        "                        if variation.variation_sequence[i] == pileup.ref_base:\n",
        "                            ALT += ref_base\n",
        "                            REF += ref_base\n",
        "                        else:\n",
        "                            condition = false\n",
        "                            pileup_file.seek(position)\n",
        "                            break\n",
        "\t\t    else: # genotype == '1/2' # OVAJ ELSE NIJE ZAVRSEN!!!!\n",
        "\t\t\t      for i in range(2):\n",
        "\t\t\t\t        variation = genotype_list[i] # u ovim slucajevima imamo samo jedan element u genotype_list.. ovako je samo lakse dalje\n",
        "            \t  if variation.variation_type == 'insertion':\n",
        "                \t  REF = pileup.ref_base\n",
        "                \t  ALT = pileup.ref_base + variation.variation_sequence\n",
        "            \t  elif variation_type == 'deletion':\n",
        "                  \tREF = pileup.ref_base + variation.variation_sequence\n",
        "                  \tALT = pileup.ref_base\n",
        "            \t  else: # SNV\n",
        "                  \tREF = pileup.ref_base\n",
        "                  \tALT = variation.variation_sequence\n",
        "        \n",
        "              \tif variation.variation_type == 'insertion' or variation.variation_type == 'deletion':\n",
        "                  \tcondition = True\n",
        "                  \twhile(condition):\n",
        "                      \tfor i in range(len(variation.variation_sequence)):\n",
        "                      \t    line = pileup_file.readline()\n",
        "                    \t      pileup = Pileup_line(line)\n",
        "                    \t      if variation.variation_sequence[i] == pileup.ref_base:\n",
        "                    \t          ALT += ref_base\n",
        "                    \t          REF += ref_base\n",
        "                      \t    else:\n",
        "                      \t        condition = false\n",
        "                      \t        pileup_file.seek(position)\n",
        "                      \t        break\n",
        "\n",
        "\n",
        "        \n",
        "    end = time.time()\n",
        "    print(end-start)"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "28.33063554763794\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}